###########################################################################################################
# Script: Define helpful functions to plot some standard population genomic statistics in sliding windows #
# Author: Frederic Labbe                                                                                  #
# Date: 12/02/2020                                                                                        #
# Affiliations: University of Notre Dame & University of Groningen                                        #
###########################################################################################################

############################################################
# Plot the population genomic statistics across the genome #

# This function plots the population genomic statistics across the genome and their lower and upper 2.5% quantiles.
# It is also possible to add the positions of a list of candidate genes on the plot using the "candidate" option.
# data: the file containing the results generated by the script popgenWindows.py.
# stat: the statistic to be considered in the calculations, i.e. "pi", "dxy", or "Fst".
# species: the species to be considered in the calculations, e.g. "Species1", "Species2", or "Species1_Species2".
# format: format of the plot, i.e. "col" (one column per scaffold), or "row" (one row per scaffold).
# plot: type of plot, i.e. "line" (line plot; default), or "scatterplot".
# candidate (optional): a list of candidate gene(s) in a file with two columns minimum: a "ID" column and a "mid" column (indicating the mid coordinates).
PlotStatWind <- function(data, stat, species, format = 'row', plot = 'line', candidate) {
  w <- max(as.numeric(eval(parse(text = paste("data$", stat, "_", species, sep = "")))))                                          # Calculate the maximum.
  x <- as.numeric(quantile(as.numeric(eval(parse(text = paste("data$", stat, "_", species, sep = "")))), 0.025, na.rm = T))       # Calculate the 2.5% quantile.
  y <- as.numeric(quantile(as.numeric(eval(parse(text = paste("data$", stat, "_", species, sep = "")))), 0.975, na.rm = T))       # Calculate the 2.5% quantile.
  if (plot == "line") {
    if (missing(candidate)) {                                                                                                       # If the list of candidate genes is not provided:
      if (format == "col") {                                                                                                        # If format is "col":
        z <- ggplot() +                                                                                                             # Use ggplot to:
          geom_line(data = data, mapping = aes(x = mid, y = as.numeric(eval(parse(text = paste(stat, "_", species, sep = "")))))) + # Plot one of the population genomic statistics across the genome.
          facet_grid(. ~ scaffold) +                                                                                                # Split the plot according to the scaffolds.
          geom_hline(yintercept = x, color = "blue", linetype = "dotted") +                                                         # Add an horizontal lines showing the 2.5% quantile.
          geom_hline(yintercept = y, color = "red", linetype = "dotted") +                                                          # Add an horizontal lines showing the 97.5% quantile.
          xlab("Chromosome position (bp)") +                                                                                        # Change the name of the x axis.
          ylab(stat) +                                                                                                              # Change the name of the y axis.
          theme_bw(base_family = "") +                                                                                              # Removed the grey base.
          #scale_y_continuous(limits = c(0, 1)) +                                                                                   # Change y axis limits.
          theme(axis.text = element_text(color = "Black"), axis.title = element_text(color = "Black"), 
                axis.title.y = element_text(hjust = 0.5), legend.title = element_blank())                                           # Remove the legend title and change color of the axis labels.
      } else {                                                                                                                      # If format is "row":
        z <- ggplot() +                                                                                                             # Use ggplot to:
          geom_line(data = data, mapping = aes(x = mid, y = as.numeric(eval(parse(text = paste(stat, "_", species, sep = "")))))) + # Plot one  of the population genomic statistics across the genome.
          facet_grid(scaffold ~ .) +                                                                                                # Split the plot according to the scaffolds.
          geom_hline(yintercept = x, color = "blue", linetype = "dotted") +                                                         # Add an horizontal lines showing the 2.5% quantile.
          geom_hline(yintercept = y, color = "red", linetype = "dotted") +                                                          # Add an horizontal lines showing the 97.5% quantile.
          xlab("Chromosome position (bp)") +                                                                                        # Change the name of the x axis.
          ylab(stat) +                                                                                                              # Change the name of the y axis.
          theme_bw(base_family = "") +                                                                                              # Removed the grey base.
          #scale_y_continuous(limits = c(0, 1)) +                                                                                   # Change y axis limits.
          theme(axis.text = element_text(color = "Black"), axis.title = element_text(color = "Black"), 
                axis.title.y = element_text(hjust = 0.5), legend.title = element_blank())                                           # Remove the legend title and change color of the axis labels.
      }
    } else {                                                                                                                          # If the list of candidate genes is provided:
      if (format == "col") {                                                                                                        # If format is "col":
        z <- ggplot() +                                                                                                             # Use ggplot to:
          geom_line(data = data, mapping = aes(x = mid, y = as.numeric(eval(parse(text = paste(stat, "_", species, sep = "")))))) + # Plot one of the population genomic statistics across the genome.
          geom_vline(data = candidate, mapping = aes(xintercept = mid), color = "black", linetype = "dotted") +                     # Add vertical line for each candidate gene.
          geom_text(data = candidate, mapping = aes(x = mid, y = w, label = ID), size = 2) +                                        # Add the name of the candidate gene(s).
          facet_grid(. ~ scaffold) +                                                                                                # Split the plot according to the scaffolds.
          geom_hline(yintercept = x, color = "blue", linetype = "dotted") +                                                         # Add an horizontal lines showing the 2.5% quantile.
          geom_hline(yintercept = y, color = "red", linetype = "dotted") +                                                          # Add an horizontal lines showing the 97.5% quantile.
          xlab("Chromosome position (bp)") +                                                                                        # Change the name of the x axis.
          ylab(stat) +                                                                                                              # Change the name of the y axis.
          theme_bw(base_family = "") +                                                                                              # Removed the grey base.
          #scale_y_continuous(limits = c(0, 1)) +                                                                                   # Change y axis limits.
          theme(axis.text = element_text(color = "Black"), axis.title = element_text(color = "Black"), 
                axis.title.y = element_text(hjust = 0.5), legend.title = element_blank())                                           # Remove the legend title and change color of the axis labels.
      } else {                                                                                                                      # If format is "row":
        z <- ggplot() +                                                                                                             # Use ggplot to:
          geom_line(data = data, mapping = aes(x = mid, y = as.numeric(eval(parse(text = paste(stat, "_", species, sep = "")))))) + # Plot one of the population genomic statistics across the genome.
          geom_vline(data = candidate, mapping = aes(xintercept = mid), color = "black", linetype = "dotted") +                     # Add vertical line for each candidate gene.
          geom_text(data = candidate, mapping = aes(x = mid, y = w, label = ID), size = 2) +                                        # Add the name of the candidate gene(s).
          facet_grid(scaffold ~ .) +                                                                                                # Split the plot according to the scaffolds.
          geom_hline(yintercept = x, color = "blue", linetype = "dotted") +                                                         # Add an horizontal lines showing the 2.5% quantile.
          geom_hline(yintercept = y, color = "red", linetype = "dotted") +                                                          # Add an horizontal lines showing the 97.5% quantile.
          xlab("Chromosome position (bp)") +                                                                                        # Change the name of the x axis.
          ylab(stat) +                                                                                                              # Change the name of the y axis.
          theme_bw(base_family = "") +                                                                                              # Removed the grey base.
          #scale_y_continuous(limits = c(0, 1)) +                                                                                   # Change y axis limits.
          theme(axis.text = element_text(color = "Black"), axis.title = element_text(color = "Black"), 
                axis.title.y = element_text(hjust = 0.5), legend.title = element_blank())                                           # Remove the legend title and change color of the axis labels.
      }
    }
  } else {
    if (missing(candidate)) {                                                                                                       # If the list of candidate genes is not provided:
      if (format == "col") {                                                                                                        # If format is "col":
        z <- ggplot() +                                                                                                             # Use ggplot to:
          geom_point(data = data, mapping = aes(x = mid, y = as.numeric(eval(parse(text = paste(stat, "_", species, sep = ""))))), color = "black", cex = 0.5) + # Plot one of the population statistics across the genome using dotes.
          facet_grid(. ~ scaffold) +                                                                                                # Split the plot according to the scaffolds.
          geom_hline(yintercept = x, color = "blue", linetype = "dotted") +                                                         # Add an horizontal lines showing the 2.5% quantile.
          geom_hline(yintercept = y, color = "red", linetype = "dotted") +                                                          # Add an horizontal lines showing the 97.5% quantile.
          xlab("Chromosome position (bp)") +                                                                                        # Change the name of the x axis.
          ylab(stat) +                                                                                                              # Change the name of the y axis.
          theme_bw(base_family = "") +                                                                                              # Removed the grey base.
          #scale_y_continuous(limits = c(0, 1)) +                                                                                   # Change y axis limits.
          theme(axis.text = element_text(color = "Black"), axis.title = element_text(color = "Black"), 
                axis.title.y = element_text(hjust = 0.5), legend.title = element_blank())                                           # Remove the legend title and change color of the axis labels.
      } else {                                                                                                                      # If format is "row":
        z <- ggplot() +                                                                                                             # Use ggplot to:
          geom_point(data = data, mapping = aes(x = mid, y = as.numeric(eval(parse(text = paste(stat, "_", species, sep = ""))))), color = "black", cex = 0.5) + # Plot one of the population statistics across the genome using dotes.
          facet_grid(scaffold ~ .) +                                                                                                # Split the plot according to the scaffolds.
          geom_hline(yintercept = x, color = "blue", linetype = "dotted") +                                                         # Add an horizontal lines showing the 2.5% quantile.
          geom_hline(yintercept = y, color = "red", linetype = "dotted") +                                                          # Add an horizontal lines showing the 97.5% quantile.
          xlab("Chromosome position (bp)") +                                                                                        # Change the name of the x axis.
          ylab(stat) +                                                                                                              # Change the name of the y axis.
          theme_bw(base_family = "") +                                                                                              # Removed the grey base.
          #scale_y_continuous(limits = c(0, 1)) +                                                                                   # Change y axis limits.
          theme(axis.text = element_text(color = "Black"), axis.title = element_text(color = "Black"), 
                axis.title.y = element_text(hjust = 0.5), legend.title = element_blank())                                           # Remove the legend title and change color of the axis labels.
      }
    } else {                                                                                                                         # If the list of candidate genes is provided:
      if (format == "col") {                                                                                                        # If format is "col":
        z <- ggplot() +                                                                                                             # Use ggplot to:
          geom_point(data = data, mapping = aes(x = mid, y = as.numeric(eval(parse(text = paste(stat, "_", species, sep = ""))))), color = "black", cex = 0.5) + # Plot one of the population statistics across the genome using dotes.
          geom_vline(data = candidate, mapping = aes(xintercept = mid), color = "black", linetype = "dotted") +                     # Add vertical line for each candidate gene.
          geom_text(data = candidate, mapping = aes(x = mid, y = w, label = ID), size = 2) +                                        # Add the name of the candidate gene(s).
          facet_grid(. ~ scaffold) +                                                                                                # Split the plot according to the scaffolds.
          geom_hline(yintercept = x, color = "blue", linetype = "dotted") +                                                         # Add an horizontal lines showing the 2.5% quantile.
          geom_hline(yintercept = y, color = "red", linetype = "dotted") +                                                          # Add an horizontal lines showing the 97.5% quantile.
          xlab("Chromosome position (bp)") +                                                                                        # Change the name of the x axis.
          ylab(stat) +                                                                                                              # Change the name of the y axis.
          theme_bw(base_family = "") +                                                                                              # Removed the grey base.
          #scale_y_continuous(limits = c(0, 1)) +                                                                                   # Change y axis limits.
          theme(axis.text = element_text(color = "Black"), axis.title = element_text(color = "Black"), 
                axis.title.y = element_text(hjust = 0.5), legend.title = element_blank())                                           # Remove the legend title and change color of the axis labels.
      } else {                                                                                                                      # If format is "row":
        z <- ggplot() +                                                                                                             # Use ggplot to:
          geom_point(data = data, mapping = aes(x = mid, y = as.numeric(eval(parse(text = paste(stat, "_", species, sep = ""))))), color = "black", cex = 0.5) + # Plot one of the population statistics across the genome using dotes.
          geom_vline(data = candidate, mapping = aes(xintercept = mid), color = "black", linetype = "dotted") +                     # Add vertical line for each candidate gene.
          geom_text(data = candidate, mapping = aes(x = mid, y = w, label = ID), size = 2) +                                        # Add the name of the candidate gene(s).
          facet_grid(scaffold ~ .) +                                                                                                # Split the plot according to the scaffolds.
          geom_hline(yintercept = x, color = "blue", linetype = "dotted") +                                                         # Add an horizontal lines showing the 2.5% quantile.
          geom_hline(yintercept = y, color = "red", linetype = "dotted") +                                                          # Add an horizontal lines showing the 97.5% quantile.
          xlab("Chromosome position (bp)") +                                                                                        # Change the name of the x axis.
          ylab(stat) +                                                                                                              # Change the name of the y axis.
          theme_bw(base_family = "") +                                                                                              # Removed the grey base.
          #scale_y_continuous(limits = c(0, 1)) +                                                                                   # Change y axis limits.
          theme(axis.text = element_text(color = "Black"), axis.title = element_text(color = "Black"), 
                axis.title.y = element_text(hjust = 0.5), legend.title = element_blank())                                           # Remove the legend title and change color of the axis labels.
      }
    }
  }
  return(z)
}

################################################################
# Extract windows having outlier population genomic statistics #

# This function extracts the windows having outlier population genomic statistics (higher than the 97.5% quantile).
# data: the file containing the results generated by the script popgenWindows.py.
# stat: the statistic to be considered in the calculations, i.e. "pi", "dxy", or "Fst".
# species: the species to be considered in the calculations, e.g. "Species1", "Species2", or "Species1_Species2".
ExtOutStat <- function(data, stat, species) {
  x <- as.numeric(quantile(as.numeric(eval(parse(text = paste ("data$", stat, "_", species, sep = "")))), 0.975, na.rm = T)) #
  y <- subset(data, as.numeric(eval(parse(text = paste (stat, "_", species, sep = "")))) > x)                                # Subset windows having pi higher than the 97.5% quantile.
  return(y)
}

